### DISCUSSION TITLE
Composer-plugin and dispatchScript

### DISCUSSION THREAD

**Discussion by Suriteka** [composer/composer#10232] [Q&A / Support]
Hello !

I created a composer package to use in my next projects.
But I would like to scope some of the dependencies used in my package to use them in my project. The script to prefix my dependencies works in my package, but when I'm installing my package in my project it doesn't launch my script.

I already know that when I install a package, the scripts of the package are not launched according to the documentation.

`Only scripts defined in the root package's composer.json are executed. If a dependency of the root package specifies its own scripts, Composer does not execute those additional scripts.`

So I found that I should create a composer-plugin. I've updated my package to be a `composer-plugin`, to launch scripts thanks to the documentation : https://getcomposer.org/doc/articles/plugins.md

I took the example of the Yoast package, they launch their prefix script during the `post-dump-autoload` as you can see here. https://github.com/Yoast/wordpress-seo/blob/trunk/composer.json (They launch static functions in their `/config/composer/actions.php` file)

But nothing is working when I install my package inside my project. I've also set `COMPOSER_DEBUG_EVENTS=1`, so as you can see the events are dispatched. I've also launch but they always returns `0` and nothing happens.

```
$event_dispatcher = $event->getComposer()->getEventDispatcher();
$io->write($event_dispatcher->dispatchScript('prefix-vendor', true));
```

![Capture d’écran 2021-10-27 à 16 13 03](https://user-images.githubusercontent.com/31373819/139083252-5823a559-e151-460c-988d-0272b607ebb8.png)

Do you have an idea from where it comes ? Is it a bad solution ?
I'm using Composer `2.1.9`
If you need more details, I'll be glad to send them.

**EDIT :** 
Here is the repo (with the branch that contains the modifications) : https://gitlab.com/bazooka/core/-/tree/2.10.0
You can look at the :
-  `composer.json` : https://gitlab.com/bazooka/core/-/blob/2.10.0/composer.json
-  `config/composer/PrefixVendorPlugin.php` : https://gitlab.com/bazooka/core/-/blob/2.10.0/config/composer/PrefixVendorPlugin.php

Thanks,

**Comment by stof** [comment]
Without seeing _where_ you put this code, there is no way we can tell you whether it can work or no

**Reply by Suriteka** [reply]
I've added them into my question! Thanks!

**Comment by stof** [comment]
Well, it looks like you expect `$event_dispatcher->dispatchScript('prefix-vendor', true)` to run the scripts defined in the composer.json of your plugin. That's not possible on purpose.
Scripts are a root-only feature, so `$event_dispatcher->dispatchScript('prefix-vendor', true)` runs the `prefix-vendor` scripts defined in the root composer.json

**Reply by Suriteka** [reply]
Thanks a lot for the reply ❤️, 

I understand it, that's what I was expecting too, but since I had `COMPOSER_DEBUG_EVENTS=1` with the `dispatch` messages working and returning `0`, I thought it was a mistake in my code.

So is there another solution to avoid the problem of multiple packages using same dependency. For example my project use a `Package A` that require `symfony/config:1.0.0` and my custom package `Core` requires `symfony/config:2.0.0`. How can I use `symfony/config:2.0.0` in my project to avoid errors. How can I prefix the dependency of my custom package to use them in my project?

**Reply by stof** [reply]
Well, even if you run your plugin on `post-autoload-dump`, I don't understand how that could help, as you would still have a dependency conflict *before* reaching that point

**Reply by Suriteka** [reply]
So there is no solutions for what I want to do? Prefixing the dependencies of my custom package when I'm requiring it?

**Reply by Seldaek** [reply]
The yoast prefixing works because yoast installs as a root package, then they build a plugin archive from that root package and prefixed dependencies, and users install that archive.

If your package is to be used as a dependency (as it should be..) then prefixing this way won't work, but it also should ideally not be necessary as Composer will be able to resolve the correct dependency version from everything in the project.
